File: clients/ios/Sources/SpectreCore/Interpolation.swift
import Foundation

/// `${...}` を含む文字列テンプレート。
///
/// 文字列全体がちょうど1つの `${...}` なら `.whole` になり、評価結果の型が保存される。
/// それ以外は `.mixed` で、各部分を文字列化して連結する (docs/spec/expression.md §1)。
public enum Template: Sendable {
    /// 補間を含まない素の文字列。
    case literal(String)
    /// 文字列全体がひとつの式。型が保存される。
    case whole(String)
    case mixed([Part])

    public enum Part: Sendable {
        case text(String)
        case expression(String)
    }
}

public enum TemplateParser {

    /// テンプレート文字列を解析する。`$${` はリテラルの `${` へのエスケープ。
    public static func parse(_ source: String) -> Template {
        guard source.contains("$") else { return .literal(source) }

        let chars = Array(source)
        var parts: [Template.Part] = []
        var text = ""
        var i = 0

        func flushText() {
            if !text.isEmpty {
                parts.append(.text(text))
                text = ""
            }
        }

        while i < chars.count {
            let c = chars[i]
            if c == "$", i + 2 < chars.count, chars[i + 1] == "$", chars[i + 2] == "{" {
                text += "${"
                i += 3
                continue
            }
            if c == "$", i + 1 < chars.count, chars[i + 1] == "{" {
                guard let end = findClosingBrace(chars, from: i + 2) else {
                    // 閉じられていない '${' はリテラルとして扱う。ここで例外にすると
                    // 「$1,000 のような文字列を書いたら画面が落ちる」ことになる。
                    text += String(chars[i...])
                    break
                }
                flushText()
                parts.append(.expression(String(chars[(i + 2)..<end])))
                i = end + 1
                continue
            }
            text.append(c)
            i += 1
        }
        flushText()

        if parts.isEmpty { return .literal("") }
        if parts.count == 1, case .expression(let source) = parts[0] { return .whole(source) }
        if parts.allSatisfy({ if case .text = $0 { return true }; return false }) {
            let joined = parts.map { part -> String in
                if case .text(let t) = part { return t }
                return ""
            }.joined()
            return .literal(joined)
        }
        return .mixed(parts)
    }

    /// `${` に対応する `}` の位置を返す。
    /// 式の中のオブジェクトリテラル `{...}` と文字列リテラル内の `}` を正しく読み飛ばす。
    private static func findClosingBrace(_ chars: [Character], from start: Int) -> Int? {
        var depth = 0
        var i = start
        while i < chars.count {
            let c = chars[i]
            if c == "'" || c == "\"" {
                i += 1
                while i < chars.count, chars[i] != c {
                    if chars[i] == "\\" { i += 1 }
                    i += 1
                }
            } else if c == "{" {
                depth += 1
            } else if c == "}" {
                if depth == 0 { return i }
                depth -= 1
            }
            i += 1
        }
        return nil
    }
}

/// テンプレートの評価。式のパースは `ExprCache` を通すため、同じ文字列の再パースは起きない。
public final class TemplateEvaluator {
    private let cache: ExprCache
    private let evaluator: ExprEvaluator
    private var templates: [String: Template] = [:]

    public init(cache: ExprCache = ExprCache(), evaluator: ExprEvaluator = ExprEvaluator()) {
        self.cache = cache
        self.evaluator = evaluator
    }

    public func template(for source: String) -> Template {
        if let cached = templates[source] { return cached }
        let parsed = TemplateParser.parse(source)
        templates[source] = parsed
        return parsed
    }

    public func evaluate(_ source: String, scope: EvalScope) -> EvalResult {
        evaluate(template(for: source), scope: scope)
    }

    public func evaluate(_ template: Template, scope: EvalScope) -> EvalResult {
        switch template {
        case .literal(let text):
            return EvalResult(.string(text))

        case .whole(let source):
            return evaluateExpression(source, scope: scope)

        case .mixed(let parts):
            var errors: [ExprError] = []
            var out = ""
            for part in parts {
                switch part {
                case .text(let text):
                    out += text
                case .expression(let source):
                    let result = evaluateExpression(source, scope: scope)
                    errors.append(contentsOf: result.errors)
                    out += result.value.stringify()
                }
            }
            return EvalResult(.string(out), errors)
        }
    }

    private func evaluateExpression(_ source: String, scope: EvalScope) -> EvalResult {
        switch cache.get(source) {
        case .success(let expr):
            return evaluator.evaluate(expr, scope: scope)
        case .failure(let error):
            return EvalResult(.null, [error])
        }
    }

    /// ドキュメント読み込み時に全式を事前解析し、パースエラーを一括検出する。
    public func precompile(_ source: String) -> [ExprError] {
        var errors: [ExprError] = []
        func check(_ exprSource: String) {
            if case .failure(let error) = cache.get(exprSource) { errors.append(error) }
        }
        switch template(for: source) {
        case .literal:
            break
        case .whole(let s):
            check(s)
        case .mixed(let parts):
            for part in parts {
                if case .expression(let s) = part { check(s) }
            }
        }
        return errors
    }

    /// この文字列が依存するスコープ相対パスの集合。差分再解決に使う。
    public func dependencies(_ source: String) -> Set<String> {
        var out = Set<String>()
        func collect(_ exprSource: String) {
            if case .success(let expr) = cache.get(exprSource) {
                out.formUnion(expr.dependencies())
            }
        }
        switch template(for: source) {
        case .literal:
            break
        case .whole(let s):
            collect(s)
        case .mixed(let parts):
            for part in parts {
                if case .expression(let s) = part { collect(s) }
            }
        }
        return out
    }
}

File: clients/android/spectre-core/src/main/kotlin/dev/spectre/core/expr/Interpolation.kt
package dev.spectre.core.expr

import dev.spectre.core.SpValue
import dev.spectre.core.stringify

/**
 * `${...}` を含む文字列テンプレート。
 *
 * 文字列全体がちょうど1つの `${...}` なら [Whole] になり、評価結果の型が保存される。
 * それ以外は [Mixed] で、各部分を文字列化して連結する (docs/spec/expression.md §1)。
 */
sealed interface Template {
    /** 補間を含まない素の文字列。 */
    data class Literal(val text: String) : Template

    /** 文字列全体がひとつの式。型が保存される。 */
    data class Whole(val source: String) : Template

    data class Mixed(val parts: List<Part>) : Template

    sealed interface Part {
        data class Text(val text: String) : Part
        data class Expression(val source: String) : Part
    }
}

object TemplateParser {

    /**
     * テンプレート文字列を解析する。
     *
     * `$${` はリテラルの `${` へのエスケープ。
     */
    fun parse(source: String): Template {
        if (!source.contains('$')) return Template.Literal(source)

        val parts = ArrayList<Template.Part>()
        val text = StringBuilder()
        var i = 0

        fun flushText() {
            if (text.isNotEmpty()) {
                parts.add(Template.Part.Text(text.toString()))
                text.clear()
            }
        }

        while (i < source.length) {
            val c = source[i]
            if (c == '$' && i + 2 < source.length && source[i + 1] == '$' && source[i + 2] == '{') {
                text.append("\${")
                i += 3
                continue
            }
            if (c == '$' && i + 1 < source.length && source[i + 1] == '{') {
                val end = findClosingBrace(source, i + 2)
                if (end < 0) {
                    // 閉じられていない '${' はリテラルとして扱う。ここで例外にすると
                    // 「$1,000 のような文字列を書いたら画面が落ちる」ことになる。
                    text.append(source.substring(i))
                    break
                }
                flushText()
                parts.add(Template.Part.Expression(source.substring(i + 2, end)))
                i = end + 1
                continue
            }
            text.append(c)
            i++
        }
        flushText()

        return when {
            parts.isEmpty() -> Template.Literal("")
            parts.size == 1 && parts[0] is Template.Part.Expression ->
                Template.Whole((parts[0] as Template.Part.Expression).source)
            parts.all { it is Template.Part.Text } ->
                Template.Literal(parts.joinToString("") { (it as Template.Part.Text).text })
            else -> Template.Mixed(parts)
        }
    }

    /**
     * `${` に対応する `}` の位置を返す。見つからなければ -1。
     *
     * 式の中のオブジェクトリテラル `{...}` と文字列リテラル内の `}` を正しく読み飛ばす。
     */
    private fun findClosingBrace(source: String, start: Int): Int {
        var depth = 0
        var i = start
        while (i < source.length) {
            when (val c = source[i]) {
                '\'', '"' -> {
                    i++
                    while (i < source.length && source[i] != c) {
                        if (source[i] == '\\') i++
                        i++
                    }
                }
                '{' -> depth++
                '}' -> {
                    if (depth == 0) return i
                    depth--
                }
            }
            i++
        }
        return -1
    }
}

/**
 * テンプレートの評価。式のパースは [ExprCache] を通すため、同じ文字列の再パースは起きない。
 */
class TemplateEvaluator(
    private val cache: ExprCache = ExprCache(),
    private val evaluator: ExprEvaluator = ExprEvaluator(),
) {
    private val templates = HashMap<String, Template>()

    fun templateOf(source: String): Template = templates.getOrPut(source) { TemplateParser.parse(source) }

    fun evaluate(source: String, scope: EvalScope): EvalResult =
        evaluate(templateOf(source), scope)

    fun evaluate(template: Template, scope: EvalScope): EvalResult = when (template) {
        is Template.Literal -> EvalResult(SpValue.Str(template.text))

        is Template.Whole -> evaluateExpression(template.source, scope)

        is Template.Mixed -> {
            val errors = ArrayList<ExprError>()
            val sb = StringBuilder()
            for (part in template.parts) {
                when (part) {
                    is Template.Part.Text -> sb.append(part.text)
                    is Template.Part.Expression -> {
                        val result = evaluateExpression(part.source, scope)
                        errors.addAll(result.errors)
                        sb.append(result.value.stringify())
                    }
                }
            }
            EvalResult(SpValue.Str(sb.toString()), errors)
        }
    }

    private fun evaluateExpression(source: String, scope: EvalScope): EvalResult =
        cache.get(source).fold(
            onSuccess = { evaluator.evaluate(it, scope) },
            onFailure = { throwable ->
                val error = (throwable as? ExprParseException)?.error
                    ?: ExprError(ExprErrorCode.E_PARSE, throwable.message ?: "式を解析できません")
                EvalResult(SpValue.Null, listOf(error))
            },
        )

    /** ドキュメント読み込み時に全式を事前解析し、パースエラーを一括検出する。 */
    fun precompile(source: String): List<ExprError> {
        val errors = ArrayList<ExprError>()
        fun check(exprSource: String) {
            cache.get(exprSource).onFailure { throwable ->
                errors.add(
                    (throwable as? ExprParseException)?.error
                        ?: ExprError(ExprErrorCode.E_PARSE, throwable.message ?: "式を解析できません")
                )
            }
        }
        when (val template = templateOf(source)) {
            is Template.Literal -> Unit
            is Template.Whole -> check(template.source)
            is Template.Mixed -> template.parts
                .filterIsInstance<Template.Part.Expression>()
                .forEach { check(it.source) }
        }
        return errors
    }

    /** この文字列が依存するスコープ相対パスの集合。差分再解決に使う。 */
    fun dependencies(source: String): Set<String> {
        val out = LinkedHashSet<String>()
        fun collect(exprSource: String) {
            cache.get(exprSource).onSuccess { out.addAll(it.dependencies()) }
        }
        when (val template = templateOf(source)) {
            is Template.Literal -> Unit
            is Template.Whole -> collect(template.source)
            is Template.Mixed -> template.parts
                .filterIsInstance<Template.Part.Expression>()
                .forEach { collect(it.source) }
        }
        return out
    }
}
